In [3]:
%matplotlib inline
import pandas as pd
import os

In [4]:
os.getcwd()

'/Users/michaelcaruana'

In [5]:
os.chdir('/Users/michaelcaruana/Desktop/pycon-pandas-tutorial')

In [6]:
from IPython.core.display import HTML
css = open('style-table.css').read() + open('style-notebook.css').read()
HTML('<style>{}</style>'.format(css))

In [10]:
titles = pd.DataFrame.from_csv('data/titles.csv', index_col=None)
titles.head()

title  year
0       Fuera de la ley  1937
1               Element  2016
2       Jeomjaengyideul  2012
3  An kubo sa kawayanan  2015
4           Hirer Angti  1992

In [8]:
cast = pd.DataFrame.from_csv('data/cast.csv', index_col=None)
cast.head()

title  year       name   type character   n
0                    Suuri illusioni  1985     Homo $  actor    Guests  22
1     Gangsta Rap: The Glockumentary  2007  Too $hort  actor   Himself NaN
2                  Menace II Society  1993  Too $hort  actor   Lew-Loc  27
3  Porndogs: The Adventures of Sadie  2009  Too $hort  actor     Bosco   3
4                 Stop Pepper Palmer  2014  Too $hort  actor   Himself NaN

### How many movies are listed in the titles dataframe?

In [18]:
#Number of movies represented by number of columns. 
titles_dedupe = titles.drop_duplicates()
titles_dedupe.shape

(217757, 2)

217,757 movies.

### What are the earliest two films listed in the titles dataframe?

In [27]:
titles_sorted = titles_dedupe.sort(['year'],ascending = True)
titles_sorted.head()

title  year
67571                                          Miss Jerry  1894
47876      Reproduction of the Corbett and Jeffries Fight  1899
77875   Pierrot's Problem, or How to Make a Fat Wife O...  1900
172960                           Trouble in Hogan's Alley  1900
39596                               Soldiers of the Cross  1900

Miss Jerry and Reproduction of Corbett adn Jeffries Fight are the first two movies

### How many movies have the title "Hamlet"?

In [29]:
hamlets = titles_dedupe[titles_dedupe.title == 'Hamlet']
hamlets.shape

(19, 2)

19 movies have the title "Hamlet"

### How many movies are titled "North by Northwest"?

In [30]:
n_by_nw = titles_dedupe[titles_dedupe.title == 'North by Northwest']
n_by_nw.shape

(1, 2)

1 movie has the title "North by Northwest"

### When was the first movie titled "Hamlet" made?

In [31]:
hamlets.year.min()

1910

1910 was when the first movie titled "Hamlet" was made.

### List all of the "Treasure Island" movies from earliest to most recent.

In [32]:
treasure_island = titles_dedupe[titles_dedupe.title == 'Treasure Island'].sort('year',ascending=True)
treasure_island.year.unique()

array([1918, 1920, 1934, 1950, 1972, 1973, 1985, 1999])

### How many movies were made in the year 1950?

In [34]:
nineteenfifty = titles_dedupe[titles_dedupe.year == 1950]
nineteenfifty.shape

(1037, 2)

In [36]:
nineteenfifty.year.unique()

array([1950])

1,037 movies were made in 1950

### How many movies were made in the year 1960?

In [38]:
nineteensixty = titles_dedupe[titles_dedupe.year == 1960]
nineteensixty.shape

(1478, 2)

1,478 movies were made in 1960

### How many movies were made from 1950 through 1959?

In [40]:
fifties = titles_dedupe[(titles_dedupe.year >= 1950) & (titles_dedupe.year <= 1959)]
fifties.shape

(12207, 2)

12,207 movies were made from 1950 through 1959.

### In what years has a movie titled "Batman" been released?

In [43]:
batman = titles_dedupe[titles_dedupe.title == 'Batman']
batman.year.unique()

array([1943, 1989])

### How many roles were there in the movie "Inception"?

In [44]:
inception = cast[cast.title == 'Inception']
len(inception)

76

In [45]:
len(cast[cast.title == 'Inception'])

76

### How many roles in the movie "Inception" are NOT ranked by an "n" value?

In [52]:
len(inception[inception.n.isnull()])

25

### But how many roles in the movie "Inception" did receive an "n" value?

In [53]:
len(inception[inception.n.notnull()])

51

### Display the cast of "North by Northwest" in their correct "n"-value order, ignoring roles that did not earn a numeric "n" value.

In [58]:
nbn_n = cast[(cast.title == 'North by Northwest') & (cast.n.notnull())].sort('n',ascending = True)
nbn_n.shape

(16, 6)

In [59]:
nbn_n.head(16)

title  year                  name     type  \
793576   North by Northwest  1959            Cary Grant    actor   
3166924  North by Northwest  1959       Eva Marie Saint  actress   
1326397  North by Northwest  1959           James Mason    actor   
2850453  North by Northwest  1959   Jessie Royce Landis  actress   
323835   North by Northwest  1959        Leo G. Carroll    actor   
2756204  North by Northwest  1959  Josephine Hutchinson  actress   
1543808  North by Northwest  1959           Philip Ober    actor   
1160305  North by Northwest  1959         Martin Landau    actor   
2225377  North by Northwest  1959         Adam Williams    actor   
1648850  North by Northwest  1959          Edward Platt    actor   
606082   North by Northwest  1959     Robert Ellenstein    actor   
2087676  North by Northwest  1959          Les Tremayne    actor   
422290   North by Northwest  1959       Philip Coolidge    actor   
1373641  North by Northwest  1959         Patrick McVey    actor   
186182   North by Northwest  1959          Edward Binns    actor   
1260364  North by Northwest  1959             Ken Lynch    actor   

                           character   n  
793576            Roger O. Thornhill   1  
3166924                  Eve Kendall   2  
1326397              Phillip Vandamm   3  
2850453              Clara Thornhill   4  
323835                 The Professor   5  
2756204                Mrs. Townsend   6  
1543808              Lester Townsend   7  
1160305                      Leonard   8  
2225377                     Valerian   9  
1648850              Victor Larrabee  10  
606082                         Licht  11  
2087676                   Auctioneer  12  
422290                     Dr. Cross  13  
1373641               Sergeant Flamm  14  
186182                Captain Junket  15  
1260364  Charley - Chicago Policeman  16

### Display the entire cast, in "n"-order, of the 1972 film "Sleuth".

In [68]:
sleuth = cast[(cast.title == 'Sleuth') & (cast.year == 1972)].sort('n',ascending = True)
sleuth.shape

(6, 6)

In [69]:
sleuth.head(9)

title  year                name     type  \
1552873  Sleuth  1972    Laurence Olivier    actor   
296388   Sleuth  1972       Michael Caine    actor   
339585   Sleuth  1972      Alec Cawthorne    actor   
1333868  Sleuth  1972  John (II) Matthews    actor   
2471035  Sleuth  1972  Eve (III) Channing  actress   
1318823  Sleuth  1972        Teddy Martin    actor   

                          character  n  
1552873                 Andrew Wyke  1  
296388                  Milo Tindle  2  
339585            Inspector Doppler  3  
1333868  Detective Sergeant Tarrant  4  
2471035             Marguerite Wyke  5  
1318823      Police Constable Higgs  6

### Now display the entire cast, in "n"-order, of the 2007 version of "Sleuth".

In [70]:
sleuth_2007 = cast[(cast.title == 'Sleuth') & (cast.year == 2007)].sort('n',ascending = True)
sleuth_2007.shape

(7, 6)

In [73]:
sleuth_2007

title  year                 name     type          character   n
296389   Sleuth  2007        Michael Caine    actor             Andrew   1
1177201  Sleuth  2007             Jude Law    actor               Milo   2
1643926  Sleuth  2007        Harold Pinter    actor        Man on T.V.   3
235239   Sleuth  2007      Kenneth Branagh    actor  Other Man on T.V. NaN
339586   Sleuth  2007  Alec (II) Cawthorne    actor  Inspector Doppler NaN
2471034  Sleuth  2007    Eve (II) Channing  actress    Marguerite Wyke NaN
3037760  Sleuth  2007    Carmel O'Sullivan  actress             Maggie NaN

### How many roles were credited in the silent 1921 version of Hamlet?

In [74]:
silent_hamlet = cast[(cast.title == "Hamlet") & (cast.year == 1921)]
len(silent_hamlet)

9

### How many roles were credited in Branagh’s 1996 Hamlet?

In [75]:
hamlet_96 = cast[(cast.title == "Hamlet") & (cast.year == 1996)]
len(hamlet_96)

54

### How many "Hamlet" roles have been listed in all film credits through history?

In [76]:
hamlet = cast[(cast.title == "Hamlet")]
len(hamlet)

314

### How many people have played an "Ophelia"?

In [78]:
ophelia = hamlet[hamlet.character == 'Ophelia']
len(ophelia)

15

In [79]:
ophelia = cast[cast.character == 'Ophelia']
len(ophelia)

100

### How many people have played a role called "The Dude"?

In [82]:
dude = cast[cast.character == 'The Dude']
dude

title  year                 name   type  \
246218                 The Big Lebowski  1998         Jeff Bridges  actor   
307990               Terms & Conditions  2015          Jordan Cann  actor   
453692                         Stranger  2000        Scott Crowell  actor   
518955           Pizza Man vs. the Dude  2004       Chris DeMarcus  actor   
555799                      Sweepstakes  1931          Mike Donlin  actor   
573038                    Self Helpless  2010       Devin The Dude  actor   
993356   Jay and Silent Bob Strike Back  2001  Matthew (XIX) James  actor   
1037738                   Explicit Ills  2008   Christopher Kadish  actor   
1244313                 American Idiots  2013     Jason Loughridge  actor   
1262052             The Winds of Autumn  1976          Steve Lyons  actor   
1377509                         The Liq  2016       Mark Alan Meek  actor   
1403264                  Wild Oats Lane  1926          Jerry Miley  actor   
1624457               Oklahoma Badlands  1948     House Peters Jr.  actor   
1673895                           Hacks  1997      Jason Priestley  actor   
2003196         Carlos Spills the Beans  2012         Bobby Styles  actor   
2124043                       Zachariah  1971      Dick Van Patten  actor   
2199681              It Can Be Arranged  2010           Joel Weiss  actor   

        character   n  
246218   The Dude   1  
307990   The Dude   9  
453692   The Dude NaN  
518955   The Dude NaN  
555799   The Dude  12  
573038   The Dude NaN  
993356   The Dude  37  
1037738  The Dude  27  
1244313  The Dude  10  
1262052  The Dude  30  
1377509  The Dude NaN  
1403264  The Dude   6  
1624457  The Dude  13  
1673895  The Dude  11  
2003196  The Dude NaN  
2124043  The Dude   9  
2199681  The Dude NaN

In [84]:
dude_actors = dude.name.drop_duplicates()
len(dude_actors)

17

### How many people have played a role called "The Stranger"?

In [85]:
stranger = cast[cast.character == 'The Stranger']
stranger_actors = stranger.name.drop_duplicates()
len(stranger_actors)

183

### How many roles has Sidney Poitier played throughout his career?

In [86]:
len(cast[cast.name == 'Sidney Poitier'])

43

### How many roles has Judi Dench played?

In [88]:
len(cast[cast.name == 'Judi Dench'])

53

### List the supporting roles (having n=2) played by Cary Grant in the 1940s, in order by year.

In [89]:
grant_support = cast[(cast.name == 'Cary Grant') & (cast.n == 2) & (cast.year >= 1940) & (cast.year <= 1949)].sort('year',ascending =True)
grant_support

title  year        name   type    character  n
793573  My Favorite Wife  1940  Cary Grant  actor         Nick  2
793583    Penny Serenade  1941  Cary Grant  actor  Roger Adams  2

### List the leading roles that Cary Grant played in the 1940s in order by year.

In [90]:
grant_lead = cast[(cast.name == 'Cary Grant') & (cast.n == 1) & (cast.year >= 1940) & (cast.year <= 1949)].sort('year',ascending =True)
grant_lead

title  year        name   type  \
793598               The Howards of Virginia  1940  Cary Grant  actor   
793555                       His Girl Friday  1940  Cary Grant  actor   
793600                The Philadelphia Story  1940  Cary Grant  actor   
793588                             Suspicion  1941  Cary Grant  actor   
793602                  The Talk of the Town  1942  Cary Grant  actor   
793579                 Once Upon a Honeymoon  1942  Cary Grant  actor   
793546                     Destination Tokyo  1943  Cary Grant  actor   
793571                             Mr. Lucky  1943  Cary Grant  actor   
793572                             Mr. Lucky  1943  Cary Grant  actor   
793580                      Once Upon a Time  1944  Cary Grant  actor   
793538                  Arsenic and Old Lace  1944  Cary Grant  actor   
793575             None But the Lonely Heart  1944  Cary Grant  actor   
793574                         Night and Day  1946  Cary Grant  actor   
793577                             Notorious  1946  Cary Grant  actor   
793594      The Bachelor and the Bobby-Soxer  1947  Cary Grant  actor   
793595                     The Bishop's Wife  1947  Cary Grant  actor   
793570  Mr. Blandings Builds His Dream House  1948  Cary Grant  actor   
793550          Every Girl Should Be Married  1948  Cary Grant  actor   
793559                I Was a Male War Bride  1949  Cary Grant  actor   

                    character  n  
793598            Matt Howard  1  
793555           Walter Burns  1  
793600     C. K. Dexter Haven  1  
793588                Johnnie  1  
793602           Leopold Dilg  1  
793579  Patrick 'Pat' O'Toole  1  
793546          Capt. Cassidy  1  
793571              Joe Adams  1  
793572        Joe Bascopolous  1  
793580            Jerry Flynn  1  
793538      Mortimer Brewster  1  
793575             Ernie Mott  1  
793574            Cole Porter  1  
793577                 Devlin  1  
793594            Dick Nugent  1  
793595                 Dudley  1  
793570          Jim Blandings  1  
793550      Dr. Madison Brown  1  
793559    Capt. Henri Rochard  1

### How many roles were available for actors in the 1950s?

In [93]:
actors_fifties = cast[(cast['type'] == 'actor') & (cast.year >= 1950) & (cast.year <=1959)]
len(actors_fifties)

151217

### How many roles were avilable for actresses in the 1950s?

In [94]:
actresses_fifties = cast[(cast['type'] == 'actress') & (cast.year >= 1950) & (cast.year <=1959)]
len(actresses_fifties)

54953

### How many leading roles (n=1) were available from the beginning of film history through 1980?

In [95]:
len(cast[(cast.n ==1)&(cast.year <=1980)])

62356

### How many non-leading roles were available through from the beginning of film history through 1980?

In [96]:
len(cast[(cast.n !=1)&(cast.year <=1980)])

1068077

### How many roles through 1980 were minor enough that they did not warrant a numeric "n" rank?

In [97]:
len(cast[(cast.n.isnull())&(cast.year <=1980)])

425953